src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,579 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	FireEtwBulkType(p_type_logger->m_n_bulk_type_value_count,
+					n_clr_instance_id,
+					i_size,
+					p_type_logger->m_p_bulk_type_event_buffer,
+					NULL,
+					NULL);
+
+	memset (p_type_logger->m_p_bulk_type_event_buffer, 0, sizeof(uint8_t) * k_size_of_event_buffer);
+	p_type_logger->m_n_bulk_type_value_count = 0;
+	p_type_logger->m_n_bulk_type_value_byte_count = 0;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// get_typeid_for_type is responsible for obtaining the unique type identifier for a
+// particular MonoType. MonoTypes are structs that are not unique pointers. There
+// can be two different MonoTypes that both System.Thread or int32 or bool[]. There
+// is exactly one MonoClass * for any type, so we leverage the MonoClass a MonoType
+// points to in order to obtain a unique type identifier in mono. With that unique
+// MonoClass, its fields this_arg and _byval_arg are unique as well.
+//
+// Arguments:
+//      * mono_type - MonoType to be logged
+//
+// Return Value:
+//      type_id - Unique type identifier of mono_type
+
+static intptr_t
+get_typeid_for_type (MonoType *t) {
+	if (m_type_is_byref (t))
+		return (intptr_t)m_class_get_this_arg (mono_class_from_mono_type_internal (t));
+	else
+		return (intptr_t)m_class_get_byval_arg (mono_class_from_mono_type_internal (t));
+}
+
+static intptr_t
+get_typeid_for_class (MonoClass *c) {
+	return get_typeid_for_type (m_class_get_byval_arg (c));
+}
+
+//---------------------------------------------------------------------------------------
+//
+// ep_rt_mono_log_single_type batches a single type into the bulk type array and flushes
+// the array to ETW if it fills up. Most interaction with the type system (type analysis)
+// is done here. This does not recursively batch up any parameter types (arrays or generics),
+// but does add their unique identifiers to the rg_type_parameters array.
+// ep_rt_mono_log_type_and_parameters is responsible for initiating any recursive calls to
+// deal with type parameters.
+//
+// Arguments:
+//      * mono_type - MonoType to be logged
+//      * type_id - Unique type identifier of mono_type
+//
+// Return Value:
+//      Index into array of where this type got batched. -1 if there was a failure.
+//
+// This follows CoreCLR's BulkTypeEventLogger::LogSingleType
+
+int
+ep_rt_mono_log_single_type (BulkTypeEventLogger *p_type_logger, MonoType *mono_type, intptr_t type_id)
+{
+	// If there's no room for another type, flush what we've got
+	if (p_type_logger->m_n_bulk_type_value_count == k_max_count_type_values)
+		ep_rt_mono_fire_bulk_type_event (p_type_logger);
+
+	EP_ASSERT (p_type_logger->m_n_bulk_type_value_count < k_max_count_type_values);
+
+	BulkTypeValue *p_val = &p_type_logger->m_rg_bulk_type_values[p_type_logger->m_n_bulk_type_value_count];
+	ep_rt_bulk_type_value_init (p_val);
```

Comment:
Since `s_name` is now `NULL`, no longer allocating or de-allocating memory for it, and renamed `ep_rt_bulk_type_value_init` to `ep_rt_bulk_type_value_clear` to keep it in line with CoreCLR's